### PR TITLE
Refresh expired azure account tokens before running query

### PR DESCRIPTION
**Pull request by cssuh**
Porting [this](https://github.com/microsoft/azuredatastudio/pull/16936) change from ADS
Before, running queries with an expired azure account token (expires after 1 hour) would fail an error message.

Now, upon running a query it will check the expiredOn field and if a token is expired, it will automatically refresh the token and run the query.

### CODE REVIEW

File: src/controllers/connectionManager.ts [hunk 1]
```diff
@@ -814,6 +818,52 @@ export default class ConnectionManager {
...
     }
 
+    public async refreshAzureAccountToken(uri: string): Promise<boolean> {
+        const profile = this.getConnectionInfo(uri);
+		if (!profile) {
+            this.vscodeWrapper.logToOutputChannel(`Connection not found for uri ${uri}`);
+			return false;
+		}
+
+        // Wait for the pending reconnction promise if any
```
Comment (Charles-Gagnon) [inline src/controllers/connectionManager.ts:828]: Spacing seems off in this file, make sure you format so everything is consistent

File: src/controllers/connectionManager.ts [hunk 2]
```diff
@@ -814,6 +819,52 @@ export default class ConnectionManager {
...
+        if (typeof expiry === 'number' && !Number.isNaN(expiry)) {
+            const currentTime = new Date().getTime() / 1000;
+            const maxTolerance = 2 * 60; // two minutes
+            if (expiry - currentTime < maxTolerance) {
+                this.vscodeWrapper.logToOutputChannel(`Access token expired for connection ${profile.connectionId} with uri ${uri}`);
+                try {
+                    let connectionResult = await this.connect(uri, profile.credentials);
+                    if (!connectionResult) {
+                        this.vscodeWrapper.showErrorMessage(`Failed to refresh connection ${profile.connectionId} with uri ${uri}, invalid connection result.`);
+                        throw new Error('Connection result is invalid');
```
Comment (Charles-Gagnon) [inline src/controllers/connectionManager.ts:855]: Maybe just `Unable to refresh connection` - "invalid" doesn't really help here

File: src/controllers/connectionManager.ts [hunk 3]
```diff
@@ -814,6 +819,52 @@ export default class ConnectionManager {
...
+                    }
+                    this.vscodeWrapper.logToOutputChannel(`Successfully refreshed token for connection ${profile.connectionId} with uri ${uri}, result: ${connectionResult}, isConnected: ${this.isConnected(uri)}, ${this.getConnectionInfo(uri)}`);
+                    return true;
+                } catch {
+                    this.vscodeWrapper.showInformationMessage('Error when refreshing token to execute query');
```
Comment (Charles-Gagnon) [inline src/controllers/connectionManager.ts:860]: This is more generic than just queries

File: src/controllers/mainController.ts [hunk 4]
```diff
@@ -734,6 +734,8 @@ export default class MainController implements vscode.Disposable {
             if (!self.connectionManager.isConnected(uri)) {
                 await self.onNewConnection();
             }
+            // check if current connection is still valid / active - if not, refresh azure account token
+            this._connectionMgr.refreshAzureAccountToken(uri);
```
Comment (Charles-Gagnon) [inline src/controllers/mainController.ts:738]: await?

File: localization/xliff/enu/constants/localizedConstants.enu.xlf [hunk 5]
```diff
@@ -329,6 +329,36 @@
         <trans-unit id="msgConnecting">
             <source xml:lang="en">Connecting to server "{0}" on document "{1}".</source>
         </trans-unit>
+        <trans-unit id="msgConnectionNotFound">
```
Comment (Charles-Gagnon) [inline localization/xliff/enu/constants/localizedConstants.enu.xlf:332]: I'm not sure about adding essentially debug messages into the output pane...but it seems fine to throw in there to make it easier to track down issues and if it gets to be too spammy later on we can change it I guess.

File: typings/vscode-mssql.d.ts [hunk 6]
```diff
@@ -5,6 +5,7 @@
 
 declare module 'vscode-mssql' {
 
+    import { Token } from '@microsoft/ads-adal-library';
```
Comment (Charles-Gagnon) [inline typings/vscode-mssql.d.ts:8]: It doesn't look like you need this anymore

File: src/controllers/connectionManager.ts [hunk 7]
```diff
@@ -814,6 +819,53 @@ export default class ConnectionManager {
         }
     }
 
+    public async refreshAzureAccountToken(uri: string): Promise<boolean> {
```
Comment (abist) [inline src/controllers/connectionManager.ts:822]: doesn't seem like the boolean is used anywhere, so the return type could just be `Promise<void>`

File: src/models/connectionProfile.ts [hunk 8]
```diff
@@ -27,6 +27,7 @@ export class ConnectionProfile extends ConnectionCredentials implements IConnect
     public emptyPasswordInput: boolean;
     public azureAuthType: AzureAuthType;
     public azureAccountToken: string;
+    public expiresOn: number;
```
Comment (abist) [inline src/models/connectionProfile.ts:30]: shouldn't this also be `number | undefined` just like how it is in `ConnectionCredentials` ?

**Review by Charles-Gagnon — APPROVED** [review]

**Review by abist — APPROVED** [review]

**Pull request merged**